pkg/globalnet/controllers/global_egressip_controller.go

```diff
@@ -194,50 +317,55 @@ func allocateIPs(key string, numberOfIPs *int, pool *ipam.IPPool, status *submar
 		return false
 	}
 
-	if *numberOfIPs == 0 {
-		tryAppendStatusCondition(&status.Conditions, &metav1.Condition{
+	if numberOfIPs == 0 {
+		tryAppendStatusCondition(&egressIP.Status.Conditions, &metav1.Condition{
 			Type:    string(submarinerv1.GlobalEgressIPAllocated),
 			Status:  metav1.ConditionFalse,
 			Reason:  "ZeroInput",
-			Message: "No global IPs to allocate",
+			Message: "The specified NumberOfIPs is 0",
 		})
-	}
 
-	if *numberOfIPs == len(status.AllocatedIPs) {
 		return false
 	}
 
-	klog.Infof("Allocating %d global IP(s) for %q", *numberOfIPs, key)
+	return true
+}
 
-	var err error
+func (c *globalEgressIPController) onRemove(numRequeues int, globalEgressIP *submarinerv1.GlobalEgressIP) bool { // nolint unparam
+	key, _ := cache.MetaNamespaceKeyFunc(globalEgressIP)
 
-	status.AllocatedIPs, err = pool.Allocate(*numberOfIPs)
-	if err != nil {
-		klog.Errorf("Error allocating IPs for %q: %v", key, err)
-		tryAppendStatusCondition(&status.Conditions, &metav1.Condition{
-			Type:    string(submarinerv1.GlobalEgressIPAllocated),
-			Status:  metav1.ConditionFalse,
-			Reason:  "IPPoolAllocationFailed",
-			Message: fmt.Sprintf("Error allocating %d global IP(s) from the pool: %v", numberOfIPs, err),
-		})
+	c.Lock()
+	defer c.Unlock()
 
+	ipSetName := c.getIPSetName(key)
+
+	if requeue := c.flushRulesAndReleaseIPs(key, numRequeues, func(allocatedIPs []string) error {
+		if globalEgressIP.Spec.PodSelector != nil {
+			return c.iptIface.RemoveEgressRulesForPods(key, ipSetName,
+				getTargetSNATIPaddress(allocatedIPs), globalNetIPTableMark)
+		} else {
+			return c.iptIface.RemoveEgressRulesForNamespace(key, ipSetName,
+				getTargetSNATIPaddress(allocatedIPs), globalNetIPTableMark)
+		}
+	}, globalEgressIP.Status.AllocatedIPs...); requeue {
 		return true
 	}
 
-	// TODO - remove IP table rules for previous allocated IPs
+	podWatcher, found := c.podWatchers[key]
+	if found {
+		close(podWatcher.stopCh)
+		delete(c.podWatchers, key)
```

Comment:
It seems we should stop the `podWatcher` even if `flushRulesAndReleaseIPs` fails above.